### PR TITLE
Optimize tracker evaluation

### PR DESCRIPTION
**Pull request by CrisBarreiro**
Task/Issue URL: https://app.asana.com/0/0/1206707308828820/f

### Description
Optimize tracker evaluation

### Steps to test this PR

_Feature 1_
- Smoke test page load

_Database migration_
- [x] Install previous version
- [x] Load wikipedia.org
- [x] Check there's an entry in app database, page_loaded_pixel_entity table
- [x] Fire button
- [x] Update the app
- [x] Check there are no entries in app database, page_loaded_pixel_entity table
- [x] Load wikipedia.org again
- [x] Check a new entry has been added

### UI changes
No UI changes

### CODE REVIEW

**Comment by CrisBarreiro** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4240?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4240** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4240?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4214** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4214?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @CrisBarreiro and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: app/src/main/java/com/duckduckgo/app/trackerdetection/TdsClient.kt [hunk 1]
```diff
@@ -49,7 +59,12 @@ class TdsClient(
         documentUrl: Uri,
         requestHeaders: Map<String, String>,
     ): Client.Result {
-        val tracker = trackers.firstOrNull { sameOrSubdomain(url, it.domain) } ?: return Client.Result(matches = false, isATracker = false)
+        val tracker = if (optimizeTrackerEvaluation) {
+            val domain = url.host?.let { Domain(it) }
+            trackers.firstOrNull { sameOrSubdomain(domain, it.domain) } ?: return Client.Result(matches = false, isATracker = false)
+        } else {
+            trackers.firstOrNull { sameOrSubdomain(url, it.domain.value) } ?: return Client.Result(matches = false, isATracker = false)
```
Comment (CrisBarreiro) [inline app/src/main/java/com/duckduckgo/app/trackerdetection/TdsClient.kt:66]: This small change doesn't affect performance. Domain is an inline value class, and therefore at runtime, instances of the inline class will be represented using the underlying type (String)

File: common/common-kotlin/src/main/java/com/duckduckgo/common/Domain.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.common
+
+@JvmInline
+value class Domain(
```
Comment (aitorvs) [inline common/common-kotlin/src/main/java/com/duckduckgo/common/Domain.kt:20]: Some comments:
* we have a `browser-api` module. I think this should rather go in there
* curious why `value class`?
* Do we need this wrapper just to wrap a string?

**Review by aitorvs — APPROVED** [review]
LGTM and tested as expected

File: common/common-kotlin/src/main/java/com/duckduckgo/common/Domain.kt (hunk 2 above)
Comment (CrisBarreiro) [reply common/common-kotlin/src/main/java/com/duckduckgo/common/Domain.kt]: This was already discussed on MM, but adding it here for transparency/traceability:

- _we have a browser-api module. I think this should rather go in there_
    - Done
- _Do we need this wrapper just to wrap a string?_
    - We don't strictly need the wrapper just to wrap a string. However, using the wrapper adds clarity and semantic meaning to the code. It also ensures that the type doesn't require the extra checks needed when directly dealing with a string, where additional validation is necessary due to uncertainty about its content.

**Pull request merged**